[metadata] Mark several functions external only

lambdageek opened the pull request:
* Marked several functions in `class.c` and `metadata.c` with `MONO_RT_EXTERNAL_ONLY` if they have `MonoError` versions that do proper error propagation instead of asserting.

* Moved `mono_class_get_method_generic` *declaration* from `metadata/` to `mini/` since it's *defintion* (and use) live in `mini`.  Also added a `MonoError*` argument and moved the assertion up to the caller.

kumpera commented on mono/dis/get.c:
```diff
@@ -1854,7 +1856,7 @@ get_field (MonoImage *m, guint32 token, MonoGenericContainer *container)
 
 static MonoGenericContainer *
-get_memberref_container (MonoImage *m, guint32 mrp_token, MonoGenericContainer *container)
+get_memberref_container (MonoImage *m, guint32 mrp_token, MonoGenericContainer *container, MonoError *error)
```
We're adding MonoError but keeping it ok when nothing is found.
This non-conforming behavior (ok and null return) must be either addressed or documented.

kumpera commented on mono/metadata/class.c:
```diff
@@ -2491,33 +2480,40 @@ mono_class_get_method_by_index (MonoClass *klass, int index)
 	for (i = 0; i < mcount; ++i) {
 		if (gklass->methods [i] == method) {
 			if (klass->methods) {
 				return klass->methods [i];
```
I believe that klass->methods can have null entries due to loading errors even though we didn't fail the class.

We should either check/fail or assert around that.

lambdageek replied:
I made a pass over `class.c` and I don't see any obvious ways that `klass->methods[i]` could ever be NULL.

Also I see a lot of places where we do something like `klass->methods[i]->....` guarded only by a check that `klass->methods != NULL`.

In any case, I'll assert that `mono_class_get_inflated_method` only returns NULL when `error` is set.

luhenry commented on mono/dis/get.c:
```diff
@@ -1864,6 +1866,7 @@ get_memberref_container (MonoImage *m, guint32 mrp_token, MonoGenericContainer *
 	guint32 table = mrp_token & 7;
 	guint32 idx = mrp_token >> 3;
 
+	error_init (error);
```
We should initialize where we create it (even without `ERROR_DECL`) since it's the way we are going towards.

lambdageek replied:
Good point.  I'll update to the same semantics as #6406 and wait for it to merge.

luhenry commented on mono/metadata/class.c:
```diff
@@ -995,7 +995,11 @@ inflate_generic_context (MonoGenericContext *context, MonoGenericContext *inflat
+	MonoError error;
+	MonoMethod *res = mono_class_inflate_generic_method_full_checked (method, NULL, context, &error);
+	if (!is_ok (&error))
+		g_error ("Could not inflate generic method due to %s", mono_error_get_message (&error));
```
Can't we use `mono_error_assert_ok` here?

lambdageek replied:
Yes, but then we would only get the message from the MonoError, without the extra text; maybe I should make an analogue of `g_assertf` for `mono_error_assert_ok`.

luhenry commented on mono/metadata/class.c:
```diff
@@ -2491,33 +2480,40 @@ mono_class_get_method_by_index (MonoClass *klass, int index)
 	int i, mcount;
 
 	g_assert (method->klass == gklass);
+	error_init (error);
```
Same as https://github.com/mono/mono/pull/6384/files#diff-f6642eb449db38819570106b8b1d484fR1869

luhenry commented on mono/metadata/class.c:
```diff
@@ -8658,8 +8657,13 @@ mono_class_get_cctor (MonoClass *klass)
+		MonoError error;
+		MonoMethod *result = mono_class_get_inflated_method (klass, mono_class_get_cctor (mono_class_get_generic_class (klass)->container_class), &error);
+		if (!is_ok (&error))
+			g_error ("Could not lookup inflated class cctor due to %s", mono_error_get_message (&error)); /* FIXME do proper error handling */
```
Same as https://github.com/mono/mono/pull/6384/files#diff-f085caaf39e861140ff40b9ef0c516d9R1001

luhenry commented on mono/mini/mini-generic-sharing.c:
```diff
@@ -477,21 +478,26 @@ mono_method_get_declaring_generic_method (MonoMethod *method)
 	MonoMethod *declaring, *m;
 	int i;
 
+	error_init (error);
```
Same as https://github.com/mono/mono/pull/6384/files#diff-f6642eb449db38819570106b8b1d484fR1869

lambdageek commented:
@kumpera:
1. Added a note in `dis/get.c` that `get_memberref_container` may return NULL on success.
2. Added an assertion in `mono_class_get_inflated_method` that it never returns NULL on success.

@luhenry:
1. Updated all new `MonoError` code to use the new pattern (initialize `error` when its declared).
2. I left the `if (!is_ok (&error)) { g_error( "explanation %s", mono_error_get_message (&error)); }` alone for now, I'll make a `mono_error_assertf_ok` PR once `g_assertf` (#6235) is merged.

luhenry commented:
@monojenkins build Linux AArch64 Interpreter

lambdageek commented:
@monojenkins build Linux AArch64
